### PR TITLE
Scrollfix

### PR DESCRIPTION
**Pull request by jamestut**
## Summary of the Pull Request

Fixed inconsistent scrolling when using both touchscreen and precision touchpad.

 
## References

#1066 and #4542

## PR Checklist
* [ ] Closes #xxx
* [X] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [X] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #4542

## Detailed Description of the Pull Request / Additional comments

Scrolling jumpiness is caused by rounding errors. Instead of retrieving the current scrolling value from `GetScrollOffset`, which is already rounded in `int`, let the scrolling operation to operate on `_scrollBar`'s `Value` directly (which uses `double`) in `TermControl.cpp`.

## Validation Steps Performed

Testing scrolling on the following scenario manually:

- nonscrollable terminal (e.g. the window is large enough to contain the current buffer).
- scrolling to the topmost and bottom-most.
- scrolling TUI apps such as `nano` and `more` in WSL.
- after clearing the terminal, both in cmd and WSL.

has the same behavior between using touchscreen or precision trackpad and regular mouse wheel.

### CODE REVIEW

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 1]
```diff
@@ -1033,19 +1033,19 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
 
             const float dy = newTouchPoint.Y - anchor.Y;
 
-            // If we've moved more than one row of text, we'll want to scroll the viewport
-            if (std::abs(dy) > fontHeight)
+            // Start viewport scroll after we've moved more than a half row of text
+            if (std::abs(dy) > (fontHeight / 2.0f))
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.cpp]: Use safemath on the division

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -1033,19 +1033,19 @@ namespace winrt::Microsoft::Terminal::TerminalControl::implementation
...
                 // Multiply by -1, because moving the touch point down will
                 // create a positive delta, but we want the viewport to move up,
                 // so we'll need a negative scroll amount (and the inverse for
                 // panning down)
                 const float numRows = -1.0f * (dy / fontHeight);
 
-                const auto currentOffset = this->GetScrollOffset();
-                const double newValue = (numRows) + (currentOffset);
+                const auto currentOffset = _scrollBar.Value();
+                const double newValue = numRows + currentOffset;
```
Comment (carlos-zamora) [inline src/cascadia/TerminalControl/TermControl.cpp]: use safemath here to force a clamped double.

**Comment by WSLUser** [comment]
https://github.com/microsoft/terminal/issues/4179 is awaiting someone to pick up before we use chromium math in the repo.

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 1 above)
Comment (jamestut) [reply src/cascadia/TerminalControl/TermControl.cpp]: Hi @carlos-zamora, I saw that the Chromium safe math headers are already in the `TermControl.cpp`'s `#include` (as part of `LibraryIncludes.h`), and I can just use those by using the `::base` namespace.
However, I am sorry for my ignorance, but why do we need to clamp this division, which is performed against a constant value (in this case 2.0f)?

Comment (carlos-zamora) [reply src/cascadia/TerminalControl/TermControl.cpp]: Sorry! Force of habit. This one probably doesn't need it...

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 2 above)
Comment (jamestut) [reply src/cascadia/TerminalControl/TermControl.cpp]: I used the chromium's safemath here for the clamped addition on the 3rd commit of this fork.

**Comment by DHowett-MSFT** [comment]
Argh: I'm really sorry about the conflict. You will need to save and commit TermControl as CRLF, then merge in master. I was not aware that we had messed up the line endings so badly. :/

**Comment by DHowett-MSFT** [comment]
FWIW: i did push a fixed version here; https://github.com/microsoft/terminal/commit/9dc594b79b123db81b684463a74cd9471df7f98a

**Comment by DHowett-MSFT** [comment]
I pushed this commit to master after doing a build test myself. Consider this merged, and congrats on your first contribution!

**Pull request closed without merging**